### PR TITLE
Enable support for project loom for Jetty's thread pool

### LINKED ISSUE #7134: Support Project Loom in Jetty 10 & JDK 17+
As of version 10, Jetty supports Project Loom and virtual threads

https://github.com/eclipse/jetty.project/issues/8007

We want to start porting our code to use Loom as soon as it goes GA.

Please add out of the box support for configuring Jetty with virtual threads in the standard HTTP Server config

### PR DESCRIPTION
**Pull request by zUniQueX**
This PR provides a new configuration option to enable virtual threads for Jetty's thread pool.

If the current runtime doesn't support virtual threads, platform threads are used instead.

### CODE REVIEW

**Comment by zUniQueX** [comment]
@dropwizard/committers This implementation should be enough to enable virtual threads. For basic testing we could ship with this implementation and add support for virtual thread executors in the `dropwizard-lifecycle` module later IMHO.

Blocked by https://github.com/dropwizard/metrics/pull/3531

File: dropwizard-core/src/main/java/io/dropwizard/core/server/AbstractServerFactory.java
Comment (mattnelson) [inline dropwizard-core/src/main/java/io/dropwizard/core/server/AbstractServerFactory.java]: I don't think jetty needs to be in the property name, it isn't in any of the other properties in this class.

Comment (zUniQueX) [reply dropwizard-core/src/main/java/io/dropwizard/core/server/AbstractServerFactory.java]: @mattnelson That was mainly to clarify only Jetty uses virtual threads at the moment. Perspectively, the managed executors will use virtual threads too.

**Comment by mattnelson** [comment]
Do any of the other thread based properties need to be handled/documented differently when virtual is enabled?

- maxThreads
- minThreads
- adminMaxThreads
- adminMinThreads
- idleThreadTimeout
- acceptorThreads
- selectorThreads

**Comment by cty123** [comment]
Thanks for working on supporting project loom, is there a plan to backport this change to release 3.x?

**Comment by zUniQueX** [comment]
Hi @cty123. As mentioned in the release notes, Dropwizard 3.x and 4.x should be kept in sync. Since the `release/4.0.x` branch is the default one, ongoing development is targeted there. After the merge the changes from 4.x will be backported to 3.x.

**Comment by zUniQueX** [comment]
@dropwizard/committers Anyone available for this PR?

**Comment by zUniQueX** [comment]
> Do any of the other thread based properties need to be handled/documented differently when virtual is enabled?
> 
>     * maxThreads
> 
>     * minThreads
> 
>     * adminMaxThreads
> 
>     * adminMinThreads
> 
>     * idleThreadTimeout
> 
>     * acceptorThreads
> 
>     * selectorThreads

minThreads, maxThreads and idleTimeout are set with the constructor of the main `InstrumentedQueuedThreadPool` and will use virtual threads now from Jetty's default thread pool.

acceptorThreads and selectorThreads obtain threads from the same thread pool and therefore will also use virtual threads.

For the admin connector the `InstrumentedQueuedThreadPool` doesn't set a `ThreadFactory` yet and therefore won't use virtual threads. I'm not sure, if applications will benefit much from virtual threads on the admin connector. But we can discuss introducing a separate property for this use case.

**Review by pstackle — APPROVED** [review]
In general, LGTM. 

I do wonder if we shouldn't fail more obviously in the case that the user configures `enableVirtualThreads: true` and it can't work (either isn't supported or the reflection code fails). It seems like Dropwizard's typical approach is to fail rather than quietly ignore a config setting  (beyond a warn log statement).

**Comment by zUniQueX** [comment]
@pstackle Thanks for the fast review!

I wanted to introduce the logs only for an easier migration path. But I agree with you, that users should be notified _more than that_, if the application cannot do what's requested. I've revised the PR and the code now throws exceptions.

I additionally added support for virtual threads on the admin connectors. Then the behavior of the connectors is more similar. But for the separate thread pool there is a separate configuration option to enable it.

Do you think we should provide any other documentation than the configuration options? I'm not sure it's a topic this big since most users will most likely not directly migrate to Java 21 and make use of virtual threads. The interested users probably have already seen the issue and this pr :wink: If you think so, I'll leave a short note in the core docs.

File: dropwizard-core/src/main/java/io/dropwizard/core/server/AbstractServerFactory.java
Comment (pstackle) [inline dropwizard-core/src/main/java/io/dropwizard/core/server/AbstractServerFactory.java]:
```suggestion
            throw new UnsupportedOperationException("Virtual threads are requested but not supported on the current runtime");
```

**Review by pstackle — APPROVED** [review]
Found one nitpick in an exception message, but otherwise, I'm happy with this.

**Comment by pstackle** [comment]
I think this feature is likely one that is worth calling out specifically in the release notes, because this is one of the main features in Java 21. I can't really think of a place in the other docs that it would make sense to reference this. Especially because we'd have to caveat it with that you must be running Java 21 or greater to even enable it.

**Comment by zUniQueX** [comment]
Sounds good :+1:

**Pull request merged**
